### DISCUSSION TITLE
How to use this library in C/C++

### DISCUSSION THREAD

**Discussion by trungnt2910** [icedland/iced#253] [Q&A] [answered]
`iced` seems to be the library with the most features available (compared to other libraries like `fcml-lib`)

However, I need to use `iced` in C/C++ code, is there any way I can link to the Rust binary from C/C++ code?

**Comment by wtfsck** [comment]
You'd have to add a C API but it would be too much work. Probably easier to just use a C/C++ library.

**Reply by trungnt2910** [reply]
I have found a few C/C++ libraries such as `fcml`, but they do not have the same functionality.

Right now I'm generating a C++/CLI wrapper for the C# version but that's _way_ too much overhead.

Are you interested in creating a C++ binding? I can help but cannot (and don't have the experience) to create and maintain such a huge library alone.

(Out of curiosity, why Rust? C/C++ is way easier to interop with Python and WASM)

**Reply by wtfsck** [reply]
> Are you interested in creating a C++ binding?

No, it's too much work.

> (Out of curiosity, why Rust? C/C++ is way easier to interop with Python and WASM)

I don't agree with this. It's super easy to compile Rust code to wasm. And creating Python modules is also very easy (basically add attributes to structs etc and the [PyO3](https://github.com/PyO3/pyo3) crate takes care of the rest).

**Reply by trungnt2910** [reply]
> > Are you interested in creating a C++ binding?
> 
> No, it's too much work.

https://raw.githubusercontent.com/icedland/iced/master/src/csharp/Intel/Iced/Intel/Assembler/Assembler.g.cs

Then can you share how big files like these are generated?

**Reply by wtfsck** [reply]
https://github.com/icedland/iced/tree/master/src/csharp/Intel/Generator/Assembler

**Reply by trungnt2910** [reply]
https://github.com/icedland/iced/blob/master/src/csharp/Intel/Generator/Enums/CSharp/CSharpEnumsGenerator.cs#L50-L180
Are _these_ generated?

**Reply by trungnt2910** [reply]
> https://github.com/icedland/iced/tree/master/src/csharp/Intel/Generator/Assembler

Also, what are the prerequisites of this generator? Some files are created automatically, while others require prior human creation. Why did you make such decisions?

**Reply by wtfsck** [reply]
Why do you think the whole lib should be auto-generated? Enums and other things can be generated, some code too but not everything.

**Reply by trungnt2910** [reply]
No I don't think the whole lib should be generated, but 100 lines of `toFullFileInfo.Add` stuff (and not using a for loop and/or reflection) _does_ seem like auto-generated.

**Comment by trungnt2910** [comment, marked as answer]
[iced_cpp](https://github.com/trungnt2910/iced_cpp)

A work-in-progress port but the first two demos, [this](https://github.com/icedland/iced/blob/master/src/csharp/Intel/README.md#disassemble-decode-and-format-instructions) and [this](https://github.com/icedland/iced/blob/master/src/csharp/Intel/README.md#assemble-instructions), works as expected.

This lays the foundation for further porting attempts as the generator seems to depend heavily on existing files.
